lib/proxy/client.go

```diff
@@ -0,0 +1,505 @@
+	}
+
+	return nil
+}
+
+// clientConn hold info about a dialed grpc connection
+type clientConn struct {
+	*grpc.ClientConn
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     *sync.WaitGroup
+
+	id   string
+	addr string
+}
+
+// Client is a peer proxy service client using grpc and tls.
+type Client struct {
+	sync.RWMutex
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	config  ClientConfig
+	conns   map[string]*clientConn
+	metrics *clientMetrics
+}
+
+type connections struct{}
+
+// NewClient creats a new peer proxy client.
+func NewClient(config ClientConfig) (*Client, error) {
+	err := config.checkAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	metrics, err := newClientMetrics()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	closeContext, cancel := context.WithCancel(config.Context)
+
+	c := &Client{
+		config:  config,
+		ctx:     closeContext,
+		cancel:  cancel,
+		conns:   make(map[string]*clientConn),
+		metrics: metrics,
+	}
+
+	if c.config.sync != nil {
+		go c.config.sync()
+	} else {
+		go c.sync()
+	}
+
+	return c, nil
+}
+
+// sync runs the proxy watcher and cleanup functionalities.
+func (c *Client) sync() {
+	ticker := c.config.Clock.NewTicker(defaults.ResyncInterval)
+	defer ticker.Stop()
+
+	proxyWatcher, err := services.NewProxyWatcher(c.ctx, services.ProxyWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component: teleport.Component(teleport.ComponentProxyPeer),
+			Client:    c.config.AccessPoint,
+			Log:       c.config.Log,
+		},
+	})
+	if err != nil {
+		c.config.Log.Errorf("Error initializing proxy peer watcher: %+v.", err)
+		return
+	}
+	defer proxyWatcher.Close()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			c.config.Log.Debug("Stopping peer proxy sync.")
+			return
+		case proxies := <-proxyWatcher.ProxiesC:
+			if err := c.updateConnections(proxies); err != nil {
+				c.config.Log.Errorf("Error syncing peer proxies: %+v.", err)
+			}
+		case <-ticker.Chan():
+			if err := c.cleanup(); err != nil {
+				c.config.Log.Errorf("Error cleaning up peer proxies: %+v.", err)
+			}
+		}
+	}
+}
+
+func (c *Client) updateConnections(proxies []types.Server) error {
+	c.Lock()
+	defer c.Unlock()
+	var errs []error
+
+	toDial := make(map[string]types.Server)
+	for _, proxy := range proxies {
+		toDial[proxy.GetName()] = proxy
+	}
+
+	for id, conn := range c.conns {
+		_, ok := toDial[id]
+
+		// delete nonexistent connections
+		if !ok {
+			c.closeConn(conn)
+			continue
+		}
+
+		// test existing connections
+		if err := c.testConnection(conn); err != nil {
+			errs = append(errs, err)
+			c.closeConn(conn)
+			continue
+		}
+	}
+
+	for id, proxy := range toDial {
+		// skips itself
+		if id == c.config.ID {
+			continue
+		}
+
+		// skip existing connection. they've been tested above.
+		if _, ok := c.conns[id]; ok {
+			continue
+		}
+
+		// establish new connections
+		conn, err := c.connect(id, proxy.GetPeerAddr())
```

Comment:
Could we avoid holding the global lock while testing/creating the connection? This would prevent calls to DialNode from getting through while we setup connections to other proxies.